### ISSUE TITLE
client_config.toml is overwritten with defaults on Apply?

### ISSUE THREAD

**Issue by Errant-4** [space-wizards/RobustToolbox#3575]
## Description

Client UI settings are not saved to disk on Apply, and ALL get overwritten with the defaults instead. This seems to impact UI, Audio and Network settings, but not keybinds.

**Reproduction**

1. IMPORTANT! Create a backup of your client_config.toml located under %appdata% or ~/.local  (...)/Space Station 14/data/
2. Connect to any server
3. Change UI scaling, click Apply
4. Close and relaunch the game - all your UI settings are reset to the defaults. (If they did not reset, try again with changing more settings)
5. Restore your backup and never touch the Options again

**Additional context**

Issue was reproduced on both windows and linux, and by multiple people. Probably most people simply do not realise the issue exists because they have no need to touch the Options anymore, but this might annoy a lot of new players who then simply did not report it

**Comment by Elijahrane** [comment]
I can confirm I've experienced this myself and received multiple complaints from it from users of all operating systems

**Comment by Errant-4** [comment]
I have not been able to reliably reproduce how the config file gets corrupted (though it definitely happens at Apply) but as a workaround, any affected player can simply restore the contents of the config file and then things will work and possibly might even be changed ingame and saved without breaking. This for example restores some common settings:

```
[audio]
mastervolume = 1.00

[display]
windowmode = 1
uiScale = 1.25

[interface]
resolutionAutoScaleUpperCutoffY = 770
resolutionAutoScaleLowerCutoffY = 240

[ui]
layout = "Separated"
```

**Comment by metalgearsloth** [comment]
Might be launcher related as this is unreproducible locally but I didn't investigate further.

**Comment by PJB3005** [comment]
The launcher itself shouldn't touch these files, so if that's the case it must be related to FULL_RELEASE or something.

**Comment by metalgearsloth** [comment]
I saw there were some toml changes a few weeks ago but I'm on master and couldn't repro it by messing with my options unless there's some specific setup for it?

**Comment by Errant-4** [comment]
I have spent some literal hours testing this, and I found that
1, if you don't have config files yet (and this is going to be the case for any new players! but you can easily test this by deleting your ~~entire config Space Station 14 dir~~ enough if you delete client_config.toml), the game loads with default config options, and when you try to apply any changes, config.toml for some reason only gets populated with
```
[interface]
resolutionAutoScaleUpperCutoffY = 770
resolutionAutoScaleLowerCutoffY = 240
```
no settings can be retained then, until you manually add the missing variable settings

2, an existing config file can get corrupted
If your config file becomes unreadable to the game, you go to 1. One way I could reproduce this is if the game crashes while in the Options (see 3), but there might be others. Anyway when this happens, the next time the game is launched from anywhere, it will fail to load the configs and launch with the defaults instead. Your config file is at this point unchanged. But when you Apply any changes to it, it will wipe out a large swathe of the config file and paste the above block in there (somewhat awkwardly, for example mine ended up like this:
```
[interface]
resolutionAutoScaleUpperCutoffY = 770
resolutionAutoScaleLowerCutoffY = 240
onAutoScaleUpperCutoffY = 770
resolutionAutoScaleLowerCutoffY = 240

[ui]
layout = "Default"
"
```
although this was possibly because my config file was not formatted as expected. In any case, after this, the configurations can not be retained again because they do not exist in the config file

3, The entire game sometimes freezes if you switch between Default and Separated UI Mode without applying. This has on at least one occasion introduced a rogue " to the end of my client_config.toml but I need to do more testing on this aspect

**Comment by PJB3005** [comment]
Oh I think I've had that happen before too. Yeah there's definitely a bug where config emits an invalid file and it then fails to load.

**Comment by ElectroJr** [comment]
Cant open PRs or test code atm, but the config man has this code snippet that I added
```cs
var cvars = _configVars.Where(x => x.Value.ConfigModified
                                   || ((x.Value.Flags & CVar.ARCHIVE) == 0 && x.Value.Value != null &&
                                       !x.Value.Value.Equals(x.Value.DefaultValue))).Select(x => x.Key);
```
which should instead have `(x.Value.Flags & CVar.ARCHIVE) != 0` and is probably related to this?

But Afaik that should just stop new modified cvars from being written, but not prevent cvars already present in the config file from changing (as it means `ConfigModified` will be true), nor should it cause them anything to be overridden by default values?

**Comment by Errant-4** [comment]
I tried changing it to != and as you suggested, it fixes being able to write changes to an empty file. Yay!

I also tried corrupting my config with a " and that still causes some strange things in the file when you Apply changes to it, and as the corrupting character of course remains, it will just fail to load the next time either. Perhaps if a config file is detected but can't be loaded, RT should rename it to get it out of the way and create a new one that it can then safely work with? In any case, this one character change still seems a massive improvement
